### PR TITLE
feat: Consume and produce docker-compose.yml via happy ingest and happy generate

### PR DESCRIPTION
**Pull request by alexlokshin-czi**
This PR extends `happy infra generate` and `happy infra ingest` to manage the docker-compose.yml file as well. Only simplistic use cases are supported (all services declared have to be built by happy, and only the `build` parameters are retained in the config file), build parameters are stored as:

```json
{
	"services": {
		"frontend": {
			"build": {
				"context": "src/api",
				"dockerfile": "Dockerfile"
			},
			"cpu": "100m"
		}
	}
}
```

It is assumed, that as soon as the management of services is done by the happy config, we are not going to treat settings retained in docker-compose.yml as the source of truth.

### CODE REVIEW

File: shared/composemanager/compose_manager.go [hunk 1]
```diff
@@ -0,0 +1,170 @@
...
+		return errors.New("unable to find services in stack config")
+	}
+
+	servicesDef := stackDef[services].(map[string]any)
+	if len(servicesDef) == 0 {
+		return errors.New("no service settings are defined in stack config")
+	}
+
+	for _, service := range c.HappyConfig.GetData().Services {
+		if sd, ok := servicesDef[service]; ok {
```
Comment (jakeyheath) [inline shared/composemanager/compose_manager.go:56]:
```suggestion
		if sd, ok := servicesDef[service]; !ok {
		  continue
		}
		
```

Comment (jakeyheath) [reply shared/composemanager/compose_manager.go:56]: https://github.com/uber-go/guide/blob/master/style.md#reduce-nesting

File: shared/composemanager/compose_manager.go [hunk 2]
```diff
@@ -0,0 +1,170 @@
...
+	if err != nil {
+		return errors.Wrap(err, "unable to get stack config")
+	}
+
+	_, ok := stackDef[services]
+	if !ok {
+		return errors.New("unable to find services in stack config")
+	}
+
+	servicesDef := stackDef[services].(map[string]any)
```
Comment (jakeyheath) [inline shared/composemanager/compose_manager.go:50]: not sure I understand the typecasting here. If you are going to typecase, handle the errors:

```suggestion
	servicesDef, ok := stackDef[services].(map[string]any)
	if !ok {
	  // return errors
	}
```

File: shared/composemanager/compose_manager.go [hunk 3]
```diff
@@ -0,0 +1,170 @@
...
+	if len(servicesDef) == 0 {
+		return errors.New("no service settings are defined in stack config")
+	}
+
+	for _, service := range c.HappyConfig.GetData().Services {
+		if sd, ok := servicesDef[service]; ok {
+			serviceDef := sd.(map[string]any)
```
Comment (jakeyheath) [inline shared/composemanager/compose_manager.go:57]: check error

File: shared/composemanager/compose_manager.go [hunk 4]
```diff
@@ -0,0 +1,170 @@
...
+			serviceConfig := types.ServiceConfig{
+				Name:     service,
+				Image:    service,
+				Profiles: []string{"*"},
+				Build:    &types.BuildConfig{},
+			}
+			platform := serviceDef[platform_architecture].(string)
```
Comment (jakeyheath) [inline shared/composemanager/compose_manager.go:64]: check error

File: shared/composemanager/compose_manager.go [hunk 5]
```diff
@@ -0,0 +1,170 @@
...
+
+			jsonData, err := json.Marshal(serviceDef[build])
+			if err != nil {
+				return errors.Wrap(err, "unable to marshal build config")
+			}
+
+			err = json.Unmarshal(jsonData, serviceConfig.Build)
+			if err != nil {
+				return errors.Wrap(err, "unable to unmarshal build config")
+			}
```
Comment (jakeyheath) [inline shared/composemanager/compose_manager.go:77]: what is this achieving?

File: shared/composemanager/compose_manager.go [hunk 6]
```diff
@@ -0,0 +1,170 @@
...
+		return errors.New("no service settings are defined in stack config")
+	}
+
+	composeServiceMap := map[string]types.ServiceConfig{}
+	for _, service := range p.Services {
+		composeServiceMap[service.Name] = service
+	}
+
+	for serviceName := range servicesDef {
+		if composeServiceDef, ok := composeServiceMap[serviceName]; ok {
```
Comment (jakeyheath) [inline shared/composemanager/compose_manager.go:146]:
```suggestion
		if composeServiceDef, ok := composeServiceMap[serviceName]; !ok {
		  continue
		}
```

Comment (jakeyheath) [reply shared/composemanager/compose_manager.go:146]: https://github.com/uber-go/guide/blob/master/style.md#reduce-nesting

File: shared/composemanager/compose_manager.go [hunk 7]
```diff
@@ -0,0 +1,170 @@
...
+			serviceDef := sd.(map[string]any)
+			serviceConfig := types.ServiceConfig{
+				Name:     service,
+				Image:    service,
+				Profiles: []string{"*"},
+				Build:    &types.BuildConfig{},
+			}
+			platform := serviceDef[platform_architecture].(string)
+			if len(platform) == 0 {
+				platform = "amd64"
```
Comment (jakeyheath) [inline shared/composemanager/compose_manager.go:66]: we should default to linux/arm64

File: shared/composemanager/compose_manager.go (hunk 1 above)
Comment (alexlokshin-czi) [reply shared/composemanager/compose_manager.go]: This is not workable, as `sd` comes out of scope, and alternative is not super elegant.

File: shared/config/happy_config.go [hunk 8]
```diff
@@ -96,6 +96,7 @@ type HappyConfig struct {
 	data *ConfigData
 
 	envConfig *Environment
+	bootstrap *Bootstrap
```
Comment (jakeyheath) [inline shared/config/happy_config.go:99]: lets not add more to these structures, pass them into the function calls. These structures are getting too messy

File: shared/composemanager/compose_manager.go (hunk 1 above)
Comment (jakeyheath) [reply shared/composemanager/compose_manager.go]: remove the nesting

File: shared/composemanager/compose_manager.go (hunk 5 above)
Comment (alexlokshin-czi) [reply shared/composemanager/compose_manager.go]: Validates whether the `build` element conforms to the service configuration build options.

File: shared/composemanager/compose_manager.go (hunk 7 above)
Comment (alexlokshin-czi) [reply shared/composemanager/compose_manager.go]: Done.

File: shared/composemanager/compose_manager.go [hunk 9]
```diff
@@ -0,0 +1,170 @@
...
+		if composeServiceDef, ok := composeServiceMap[serviceName]; ok {
+			serviceDef := servicesDef[serviceName].(map[string]any)
+			serviceDef[build] = composeServiceDef.Build
+
+			composePlatformArchitecture := "linux/amd64"
+			if len(composeServiceDef.Platform) > 0 {
+				composePlatformArchitecture = composeServiceDef.Platform
+			}
+
+			platformArchitecture := "amd64"
```
[the remark this answers is not in the document]
Comment (alexlokshin-czi) [reply shared/composemanager/compose_manager.go]: Done.

**Review by jakeyheath — APPROVED** [review]

File: shared/config/happy_config.go (hunk 8 above)
Comment (alexlokshin-czi) [reply shared/config/happy_config.go:99]: I think this one was missing from the happy config from the very beginning. This is essential contextual information that was never made available to happy. I am not in support of passing bootstrap references separate from HappyConfig, when HappyConfig is instantiated as `config.NewHappyConfig(bootstrapConfig)`; bootstrap has no meaning outside of HappyConfig.

File: shared/composemanager/compose_manager.go (hunk 1 above)
Comment (alexlokshin-czi) [reply shared/composemanager/compose_manager.go]: Done

File: shared/composemanager/compose_manager.go (hunk 5 above)
Comment (alexlokshin-czi) [reply shared/composemanager/compose_manager.go]: This is not free-form data, as it has to conform with the service build config on the way out.

**Review by jakeyheath — APPROVED** [review]

**Pull request merged**